File: segmentManager/src/main/java/com/clumob/segment/manager/Segment.java
package com.clumob.segment.manager;

import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.LifecycleRegistry;
import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;

import androidx.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.clumob.log.AppLog;
import com.clumob.segment.controller.SegmentController;
import com.clumob.segment.controller.SegmentInfo;
import com.clumob.segment.controller.Storable;


/**
 * Created by prashant.rathore on 02/02/18.
 */

public class Segment<VM, Controller extends SegmentController<VM>> implements LifecycleOwner {


    LifecycleRegistry mLifecycleRegistry = new LifecycleRegistry(this);

    @NonNull
    @Override
    public Lifecycle getLifecycle() {
        return mLifecycleRegistry;
    }

    public enum SegmentState {
        FRESH,
        CREATE,
        START,
        RESUME,
        PAUSE,
        STOP,
        DESTROY
    }

    private final Controller controller;
    private final SegmentViewHolderFactory screenFactory;

    private SegmentViewHolder<VM, Controller> boundedView = null;
    private SegmentInfo<Storable, Storable> segmentInfo;

    private Context context;
    private LayoutInflater layoutInflater;

    SegmentState currentState = SegmentState.FRESH;

    public Segment(SegmentInfo segmentInfo, Controller controller, SegmentViewHolderFactory screenFactory) {
        this.screenFactory = screenFactory;
        this.segmentInfo = segmentInfo;
        this.controller = controller;
    }

    public void attach(Context context, LayoutInflater layoutInflater) {
        this.context = context;
        this.layoutInflater = layoutInflater;
    }

    public SegmentInfo getSegmentInfo() {
        return segmentInfo;
    }

    public SegmentViewHolder createView(ViewGroup parentView) {
        SegmentViewHolder<?, ?> segmentViewHolder = screenFactory.create(context, layoutInflater, parentView);
        return segmentViewHolder;
    }

    public SegmentViewHolder getBoundedView() {
        return boundedView;
    }

    public void onCreate() {
        switch (this.currentState) {
            case FRESH:
            case DESTROY:
                createInternal();
                break;
        }
    }

    private void createInternal() {
        currentState = SegmentState.CREATE;
        controller.onCreate();
        controller.restoreState(segmentInfo.getRestorableSetmentState());
        mLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_CREATE);
    }

    public void bindView(SegmentViewHolder<VM, Controller> viewHolder) {
        boundedView = viewHolder;
        boundedView.attachLifecycleOwner(this);
        boundedView.bind(this, controller.getViewData(), controller);
//        boundedView.restoreState(segmentInfo.getSavedViewState());
    }

    public void onConfigurationChanged(Configuration newConfig) {
        boundedView.onConfigurationChanged(newConfig);
    }

    public void onStart() {
        switch (currentState) {
            case DESTROY:
            case FRESH:
            case CREATE:
                onCreate();
            case STOP:
                startInternal();
            case START:
                break;
            case RESUME:
                break;
            case PAUSE:
                break;
        }
    }

    private void startInternal() {
        currentState = SegmentState.START;
        controller.onStart();
        boundedView.onStart();
        mLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_START);
    }

    public void onResume() {
        if (currentState != SegmentState.RESUME) {
            onStart();
            resumeInternal();
        }
    }

    private void resumeInternal() {
        currentState = SegmentState.RESUME;
        boundedView.resume();
        controller.onResume();
        mLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_RESUME);
    }

    public void onPause() {
        switch (currentState) {
            case FRESH:
            case CREATE:
                onStart();
            case RESUME:
                pauseInternal();
                break;
            case START:
            case PAUSE:
            case STOP:
            case DESTROY:
                break;
        }
    }

    private void pauseInternal() {
        currentState = SegmentState.PAUSE;
        mLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_PAUSE);
        controller.onPause();
        boundedView.pause();
        final Storable viewState = boundedView.createStateSnapshot();
        segmentInfo.setRestorableSegmentState(viewState);

    }

    public void onStop() {
        switch (this.currentState) {
            case DESTROY:
            case FRESH:
                onCreate();
                break;
            case RESUME:
                onPause();
            case PAUSE:
            case START:
                stopInternal();
                break;
        }
    }


    private void stopInternal() {
        currentState = SegmentState.STOP;
        mLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_STOP);
        boundedView.onStop();
        controller.onStop();

    }

    public void unBindView() {
        if (boundedView != null) {
            boundedView.detachLifecycleOwner();
            boundedView.unBind();
            boundedView = null;
        }
    }

    public void onDestroy() {
        if (this.currentState != SegmentState.DESTROY) {
            onStop();
            unBindView();
            destroyInternal();
        }
    }

    private void destroyInternal() {
        currentState = SegmentState.DESTROY;
        mLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_DESTROY);
        controller.onDestroy();
    }

    void dettach() {
        this.context = null;
        this.layoutInflater = null;
    }

    public void onActivityResult(int code, int resultCode, Intent data) {
        boundedView.onActivityResult(code, resultCode, data);
    }

    public void onRequestPermissionsResult(int code, String[] permissions, int[] grantResults) {
        boundedView.onRequestPermissionsResult(code, permissions, grantResults);
    }

    public boolean handleBackPressed() {
        if(boundedView == null) {
            AppLog.d("SEGMENT","SegmentInfo " + segmentInfo);
            AppLog.printStack(new NullPointerException("Cannot handle backpressed SegmentView is Null"));
            return false;
        }
        else {
            return boundedView.handleBackPressed();
        }

    }

    public boolean isResumed() {
        return currentState == SegmentState.RESUME;
    }

}

File: segmentManager/src/main/java/com/clumob/segment/manager/Segment.kt
package com.clumob.segment.manager

import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.LifecycleRegistry
import com.clumob.log.AppLog
import com.clumob.segment.controller.SegmentInfo
import com.clumob.segment.controller.common.Controller
import com.clumob.segment.view.SegmentViewProvider

/**
 * Created by prashant.rathore on 02/02/18.
 */
open class Segment(protected val controller: Controller,
                   protected val screenFactory: SegmentViewProvider) : LifecycleOwner {

    var mLifecycleRegistry = LifecycleRegistry(this)

    override fun getLifecycle(): Lifecycle {
        return mLifecycleRegistry
    }

    enum class SegmentState {
        FRESH, CREATE, START, RESUME, PAUSE, STOP, DESTROY
    }

    private var boundedView: SegmentViewHolder? = null
    private lateinit var segmentInfo: SegmentInfo
    private var isBounded = false
    private var context: Context? = null
    private var layoutInflater: LayoutInflater? = null
    var currentState = SegmentState.FRESH

    fun bindSegmentInfo(segmentInfo: SegmentInfo) {
        if (isBounded && segmentInfo != this.segmentInfo)
            throw IllegalAccessException("SegmentInfo already bound. Cannot be changed")
        this.segmentInfo = segmentInfo
        isBounded = true
    }

    fun attach(context: Context, layoutInflater: LayoutInflater) {
        this.context = context
        this.layoutInflater = layoutInflater
    }

    fun getSegmentInfo(): SegmentInfo {
        return segmentInfo
    }

    fun createView(parentView: ViewGroup?): SegmentViewHolder {
        return screenFactory.create(parentView, controller.getType())
    }

    fun getBoundedView(): SegmentViewHolder? {
        return boundedView
    }

    fun onCreate() {
        when (currentState) {
            SegmentState.FRESH, SegmentState.DESTROY -> createInternal()
        }
    }

    private fun createInternal() {
        currentState = SegmentState.CREATE
        controller.onCreate()
        controller.onRestore(segmentInfo.restorableSetmentState)
        mLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_CREATE)
    }

    fun bindView(viewHolder: SegmentViewHolder) {
        boundedView = viewHolder as SegmentViewHolder
        boundedView!!.attachLifecycleOwner(this)
        boundedView!!.bind(controller)
        //        boundedView.restoreState(segmentInfo.getSavedViewState());
    }

    fun onConfigurationChanged(newConfig: Configuration?) {
        boundedView!!.onConfigurationChanged(newConfig)
    }

    fun onStart() {
        when (currentState) {
            SegmentState.DESTROY, SegmentState.FRESH, SegmentState.CREATE -> {
                onCreate()
                startInternal()
            }
            SegmentState.STOP -> startInternal()
            SegmentState.START -> {
            }
            SegmentState.RESUME -> {
            }
            SegmentState.PAUSE -> {
            }
        }
    }

    private fun startInternal() {
        currentState = SegmentState.START
        controller.onStart()
        boundedView!!.onStart()
        mLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_START)
    }

    fun onResume() {
        if (currentState != SegmentState.RESUME) {
            onStart()
            resumeInternal()
        }
    }

    private fun resumeInternal() {
        currentState = SegmentState.RESUME
        boundedView!!.resume()
        controller.onResume()
        mLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_RESUME)
    }

    fun onPause() {
        when (currentState) {
            SegmentState.FRESH, SegmentState.CREATE -> {
                onStart()
                pauseInternal()
            }
            SegmentState.RESUME -> pauseInternal()
            SegmentState.START, SegmentState.PAUSE, SegmentState.STOP, SegmentState.DESTROY -> {
            }
        }
    }

    private fun pauseInternal() {
        currentState = SegmentState.PAUSE
        mLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_PAUSE)
        controller.onPause()
        boundedView!!.pause()
        val viewState = boundedView!!.createStateSnapshot()
        segmentInfo.setRestorableSegmentState(viewState)
    }

    fun onStop() {
        when (currentState) {
            SegmentState.DESTROY, SegmentState.FRESH -> onCreate()
            SegmentState.RESUME -> {
                onPause()
                stopInternal()
            }
            SegmentState.PAUSE, SegmentState.START -> stopInternal()
        }
    }

    private fun stopInternal() {
        currentState = SegmentState.STOP
        mLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_STOP)
        boundedView!!.onStop()
        controller.onStop()
    }

    fun unBindView() {
        if (boundedView != null) {
            boundedView!!.detachLifecycleOwner()
            boundedView!!.unBind()
            boundedView = null
        }
    }

    fun onDestroy() {
        if (currentState != SegmentState.DESTROY) {
            onStop()
            destroyInternal()
            unBindView()
        }
    }

    private fun destroyInternal() {
        currentState = SegmentState.DESTROY
        mLifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_DESTROY)
        controller.onDestroy()
    }

    fun dettach() {
        context = null
        layoutInflater = null
    }

    fun onActivityResult(code: Int, resultCode: Int, data: Intent?) {
        boundedView!!.onActivityResult(code, resultCode, data)
    }

    fun onRequestPermissionsResult(code: Int, permissions: Array<String>, grantResults: IntArray?) {
        boundedView!!.onRequestPermissionsResult(code, permissions, grantResults!!)
    }

    fun handleBackPressed(): Boolean {
        return if (boundedView == null) {
            AppLog.d("SEGMENT", "SegmentInfo $segmentInfo")
            AppLog.printStack(NullPointerException("Cannot handle backpressed SegmentView is Null"))
            false
        } else {
            boundedView!!.handleBackPressed()
        }
    }

    val isResumed: Boolean
        get() = currentState == SegmentState.RESUME


}